### PR TITLE
hw6 ml Чеснов Антон

### PR DESCRIPTION
**Pull request by ChesnovAE**
Loss: MAE, optim: ADAGRAD, regularization: Elastic

### CODE REVIEW

**Comment by krabye** [comment]
Homework 6 ml soft deadline.

**Comment by krabye** [comment]
Homework 6 ml hard deadline.

File: homeworks/homework_06_ml/linear_regression.py [hunk 1]
```diff
@@ -1,17 +1,36 @@
...
         :param lambda_coef: constant coef for gradient descent step
-        :param regulatization: regularizarion type ("L1" or "L2") or None
+        :param regulatization: regularizarion type ("L1" or "L2" or "elastic") or None
         :param alpha: regularizarion coefficent
         :param batch_size: num sample per one model parameters update
         :param max_iter: maximum number of parameters updates
         """
-        raise NotImplementedError
+        self.lambda_coef = lambda_coef
+        self.regulatization = regulatization
```
Comment (krabye) [inline homeworks/homework_06_ml/linear_regression.py:23]: Неплохо бы проверить, что входной параметр соответствует реализации. Лучше в ините, чем в середине алгоритма, когда память выделилась и тп)

File: homeworks/homework_06_ml/linear_regression.py [hunk 2]
```diff
@@ -20,19 +39,74 @@ def fit(self, X_train, y_train):
...
+        # добавляем фиктивный единичный столбец
+        y_train = y_train.reshape(-1, 1)
+        X_train = np.hstack((np.ones((X_train.shape[0], 1)), X_train))  # X_train.shape = (n, m + 1)
+        self._w = np.random.normal(scale=0.001, size=(1, X_train.shape[1]))  # self._w.shape = (m + 1, 1)
+        g = np.zeros_like(self._w)
+        eps = 1e-7
+        data = np.hstack((X_train, y_train))
+        for i in range(self.max_iter):
+            n_batches = X_train.shape[0] // self.batch_size  # считаем количество целый батчей в X_train
+            data_perm = np.random.permutation(data)  # рандомно мешаем data
```
Comment (krabye) [inline homeworks/homework_06_ml/linear_regression.py:53]: Тут происходит копия памяти, можно было просто завести массив индексов и его перемешивать каждый раз.

File: homeworks/homework_06_ml/linear_regression.py [hunk 3]
```diff
@@ -20,19 +39,74 @@ def fit(self, X_train, y_train):
...
         """
-        pass
+        return self._w
+
+    def step(self, w, x, y_true):
+        gradient = np.zeros_like(w)
+        for j in range(x.shape[1]):
+            sum_ = 0
+            for i in range(x.shape[0]):
+                sum_ += x[i, j] * np.sign(x[i, :] @ w.T - y_true[i])
```
Comment (krabye) [inline homeworks/homework_06_ml/linear_regression.py:96]: Матричные операции требовали без циклов, минус балл.

File: homeworks/homework_06_ml/linear_regression.py [hunk 4]
```diff
@@ -20,19 +39,74 @@ def fit(self, X_train, y_train):
...
+            y_train = data_perm[:, -1].reshape(-1, 1)  # y_train.shape = (n, 1)
+            loss = 0
+            for idx in range(n_batches):
+                sample = X_train[idx * self.batch_size:(idx + 1) * self.batch_size]
+                y_true = y_train[idx * self.batch_size:(idx + 1) * self.batch_size]
+                y_pred = self.linear(self._w, sample)
+                loss += self.mae(y_true, y_pred)
+                gradient = self.step(self._w, sample, y_true)
+                g += gradient ** 2
+                self._w += - self.lambda_coef / self.batch_size * gradient / np.sqrt(g + eps)
```
Comment (krabye) [inline homeworks/homework_06_ml/linear_regression.py:64]: такое себе замечание: можно было бы изначально инициализировать g=eps и потом не прибавлять eps каждый раз

File: homeworks/homework_06_ml/linear_regression.py (hunk 1 above)
Comment (TheTonyPub) [reply homeworks/homework_06_ml/linear_regression.py:23]: Но это в текущей задаче. По идее, если делать класс заранее универсальным, то лучше проверять там, где делается регуляризация, чтобы можно было потом допилить ещё две регуляризации и все работало.

File: homeworks/homework_06_ml/linear_regression.py (hunk 2 above)
Comment (TheTonyPub) [reply homeworks/homework_06_ml/linear_regression.py:53]: Да, уже после подумал, что так лучше)

File: homeworks/homework_06_ml/linear_regression.py (hunk 4 above)
Comment (TheTonyPub) [reply homeworks/homework_06_ml/linear_regression.py:64]: А, ну да, можно и так :)

File: homeworks/homework_06_ml/linear_regression.py (hunk 3 above)
Comment (TheTonyPub) [reply homeworks/homework_06_ml/linear_regression.py:96]: Справедливо) Просто в голове и на листке в тот раз не складывались все размерности и представления, поэтому решил с потерей баллов, но чтобы наверняка работало :Р
